hbase-zookeeper/src/main/java/org/apache/hadoop/hbase/zookeeper/ZKUtil.java

```diff
@@ -741,6 +741,10 @@ public static int getNumberOfChildren(ZKWatcher zkw, String znode)
     if (nodes != null) {
       List<NodeAndData> newNodes = new ArrayList<>();
       for (String node : nodes) {
+        if (Thread.interrupted()) {
+          // Partial data should not be processed. Cancel processing by sending empty list.
```

Comment:
@arshadmohammad have you seen this issue being the cause for create table being slow or this is an improvement over preempting any existing nodeChildrenChanged event processing (done in ZKPermissionWatcher)? 
Basically, when thread is interrupted and we return empty list, what could be the catch? Some incorrect decision could be taken by HMaster?

Reply:
For branch-1.3.x this is an issue, the reason I have explained in the issue description. 
For master and branch-2, this PR change is for improvement as these branches do not use zk based region assignment.

Code change is same, for branch-1.3.x  it is for the fix for other branches it is just improvement.

Reply:
getChildDataAndWatchForNewChildren is called from many places, but it is interrupted from ZKPermissionWatcher#nodeChildrenChanged() by calling childrenChangedFuture.cancel.  Here cancel/interrupt  indicates that we have more updated data, no need to proceed with the old data. Returning empty is for graceful exit of the processing.

Reply:
Thanks @arshadmohammad 
It seems `ZKPermissionWatcher#nodeChildrenChanged()` is interrupting `ZKPermissionWatcher#refreshNodes()` and that method is already taking care of `Thread.interrupted()` check. That would not interrupt `ZKUtil#getChildDataAndWatchForNewChildren()`.

Reply:
@virajjasani for the reviews
I have not understood if there is any concern on the change from your comment. But still let me explain bit more.

When cancel is called on childrenChangedFuture, irrespective of which method is executing inside run method, run method should finish. Inside run method we are invoking two independent methods getChildDataAndWatchForNewChildren and refreshNodes. So both should handle the interrupted state of the thread. refreshNodes method was already handling it, so i have added handling only for getChildDataAndWatchForNewChildren.

Reply:
Oh okk, while looking into `asyncProcessNodeUpdate()`, I was actually referring to the current code and not your patch, hence I didn't find call stack from ZKPermissionWatcher -> ZKUtil#getChildDataAndWatchForNewChildren().
Got it, all good.
Thanks